### DISCUSSION TITLE
Varying results with `pagedown::chrome_print()`

### DISCUSSION THREAD

**Discussion by snhansen** [quarto-dev/quarto-cli#11174] [Q&A]
### Description

I'm using RevealJS html-slides created with Quarto more and more, and to create pdf handouts I've been using `pagedown::chrome_print()` which works very well. I have however noticed a very random bug when converting my slides: the pdf produced by `pagedown::chrome_print()` is not always exactly the same. Most of the times, the conversion is flawless but some times all LaTeX formulas are missing while other times some but not all formulas are missing. It appears more frequently with large decks, so it's taken me some time to create a minimal reproducible example. I'm not sure whether this is a Quarto or pagedown/chromote thing, but I'd be happy to see if anyone other than me can recreate the problem and to get some input on what might be causing this.

Consider this `slides.qmd`:

````
---
format: 
  revealjs:
    slide-number: true
    smaller: true
---

## The average

Let $y_1,\ldots,y_n$ be a sample of size $n$. The average is defined as

$$
\bar{y} = \frac{1}{n}\sum_{i=1}^n y_i = (y_1+y_2+\cdots + y_n)/n
$$


```{r}
#| echo: false
#| out-width: 600px
ggplot2::ggplot(data = mtcars, ggplot2::aes(x = mpg, y = disp)) + ggplot2::geom_point()
```
````

I've used the following script to create 350 pdf's of the html slides

``` r
convert <- function(i) {
  pagedown::chrome_print(input = "slides.html", output = paste0("slides-", stringr::str_pad(i, width = 4, pad = "0"), ".pdf"))
}

quarto::quarto_render("slides.qmd", cache_refresh = TRUE)
purrr::map(1:350, convert)
```

When I look at their sizes I see this pattern:

``` r
pdf_sizes
63604 68842 69374 
    6   340     4 
```

The majority of the slides look like this (which matches the html):
![image](https://github.com/user-attachments/assets/e99a37ef-062e-47f9-a139-1fb74e3d8bfd)

The larger files are because the plot is pushed to its own slide which is weird because everything else is unchanged. The smaller files are because some of the LaTeX is stripped like this:
![image](https://github.com/user-attachments/assets/c7931972-0db3-4951-852a-4ebe7ffb2c2e)

The above has been created using

* Windows 11
* Quarto 1.6.28
* R 4.4.1
* pagedown 0.21

but the issue has been bugging me for quite a while, so it's definitely not something new. Any ideas what's at stake here?

**Comment by mcanouil** [comment]
Did you try setting some wait time before the screenshots happen? It might simply be because MathJax was not fully loaded which means the math layout is not finished.

**Reply by snhansen** [reply]
> Did you try setting some wait time before the screenshots happen? It might simply be because MathJax was not fully loaded which means the math layout is not finished.

Yes, I have experimented with that argument. I've just tried the example above with `wait = 200` and can still reproduce the issue.

**Reply by mcanouil** [reply]
What's the unit?

(I'm not sure this is a Quarto issue here)

**Reply by snhansen** [reply]
> What's the unit?
> 
> (I'm not sure this is a Quarto issue here)

It's in seconds. The default is 2. Not sure it's a Quarto issue either, but maybe someone else using the Quarto + pagedown combo has experienced similar issues which is why I've posted here.

**Comment by cderv** [comment]
For revealjs printing to PDF, I found I had the best results with using decktape to make the PDF printing. 
This is discussed in 
- https://github.com/quarto-dev/quarto-cli/issues/4677

and linked issues 

I don't have ideas on what could cause the Math to not be fully rendered 🤔 `chrome_print()` has a specific JS script to try to sync things up before making the printing and maybe it needs adjustement to latest reveal version. Script is here: https://github.com/rstudio/pagedown/blob/main/inst/resources/js/chrome_print.js
There is a wait for Mathjax there. 

You could also check what Revealjs gives you before printing in browser by looking at the [print-pdf version](https://quarto.org/docs/presentations/revealjs/presenting.html#print-to-pdf) of the slide. This would make sure there is nothing weird already there before triggering a chrome print. 

All this PDF printing is touchy, with lots of element. I would really try out other printing tools so that we can know if this is an issue with **pagedown** itself, or other layers. 

Hope it helps

**Reply by snhansen** [reply]
Thanks for chiming in, @cderv, and thanks for the suggestions!

I previously tried decktape and decided not to use it because it doesn't print diagrams made with mermaid very well. I might consider using something else than mermaid since that's the only issue with decktape I've encountered (apart from the issue that you have to manually set the slides to render via the `--slides` argument otherwise it never stops).

I tried the looking at the `print-pdf-version` (for another set of slides) and most of the times the math is rendered correctly but other times the math isn't rendered at all:
![image](https://github.com/user-attachments/assets/14b6e05c-eae5-4905-b8bf-823806139f09)

I guess this is what happens when `pagedown` fails. Does this mean it's a RevealJS issue?

**Reply by mcanouil** [reply]
What about the R package `webshot2`?

Roughly something like (plus some other tool to aggregate the images into a PDF):

```r
max_slide=12
for (islide in seq_len(max_slide)) {
  webshot2::webshot(
    url = paste0("revealjs-slides.html#", islide),
    file = sprintf("slide-%02d.png", islide),
    vwidth = 1920,
    vheight = 1080
   )
}
```

**Reply by cderv** [reply]
> and most of the times the math is rendered correctly but other times the math isn't rendered at all:
> (...)
> I guess this is what happens when pagedown fails. Does this mean it's a RevealJS issue?

Possibly yes, or something related to Pandoc integration. Is this presentation with numerous slide ? does it happens after some slides ? 

There are some pre-loading happening when navigating a presentation, and maybe in pdf print view something is not working 🤔 

I would be interested with an issue and a reproducible example if you have one

**Reply by snhansen** [reply]
> What about the R package `webshot2`?

That's a good suggestion. I might try it out!

> Possibly yes, or something related to Pandoc integration. Is this presentation with numerous slide ? does it happens after some slides ?
> 
> There are some pre-loading happening when navigating a presentation, and maybe in pdf print view something is not working 🤔
> 
> I would be interested with an issue and a reproducible example if you have one

Yes, this is a presentation with a fair amount of slides. I can reproduce the issue fairly easily with several of my decks which all contain numerous slides. When the issue happens it's all math throughout the presentation that isn't rendered. I haven't been able to reproduce the issue with `print-pdf` with the minimal example in the initial post yet. 

One thing I did notice however is that for some slides pressing "E" yields `slides.html?print-pdf=#/...` while for others it yields `slides.html?view=print#/...` Is there a difference between the two?

And by the way, I tested this `html-math-method: katex` and can't reproduce the issue with that, so it seems connected to MathJax.

**Reply by snhansen** [reply]
Alright, I just reproduced it with a "minimal" example: I took the slide from the intial post and created a deck of 50 identical slides (not sure this was necessary but my feeling is that the "heavier" the presentation the more the issue presents itself). In my browser I went back and forth between `print-pdf-mode` using "E" until I saw the issue (had to do this quite a lot so it's hard to reproduce). Below is the result:

![image](https://github.com/user-attachments/assets/56ea98cf-0210-42ab-a845-fbc7fb54f6a5)

Note how it's only one math block that isn't rendered at all (neither code nor output) while in my other example it was all math that wasn't rendered but their LaTeX code was still here.

**Reply by mcanouil** [reply]
This seems to be a Reveal.js issue.

**Reply by cderv** [reply]
> One thing I did notice however is that for some slides pressing "E" yields slides.html?print-pdf=#/... while for others it yields slides.html?view=print#/... Is there a difference between the two?

The second URL is from Revealjs v5 update that we included in Quarto 1.6. So you may not be rendering the presentation with the same version. Though both url are fine and supported by revealjs. They introduced the second one because of new `view=scroll` feature. 

> And by the way, I tested this html-math-method: katex and can't reproduce the issue with that, so it seems connected to MathJax.
> (...)
> This seems to be a Reveal.js issue.

That is really good to know. Could be a Mathjax issue or Revealjs issue integration with Mathjax 🤔 

To be sure, it would be great to reproduce without all Quarto doing. One easy way would be to take the .qmd, remove everything that makes is a .qmd to keep only what is supported by pandoc, then make it a .md. From there, call `pandoc --to reveal --mathjax ...` with all command line that are needed. 

If the issue reproduce, then this is definitely not Quarto !

**Reply by snhansen** [reply]
> To be sure, it would be great to reproduce without all Quarto doing. One easy way would be to take the .qmd, remove everything that makes is a .qmd to keep only what is supported by pandoc, then make it a .md. From there, call `pandoc --to reveal --mathjax ...` with all command line that are needed.
> 
> If the issue reproduce, then this is definitely not Quarto !

I'll give it a try, thanks. 

Edit: I can't seem to reproduce the error this way.. weird!

**Reply by cderv** [reply]
> I can't seem to reproduce the error this way.. weird!

Not so weird. It could mean this is one of the configuration that Quarto is doing. 

Either configuration we set as default, which is not the default for revealjs in Pandoc, or it could be one of the specific processing Quarto is doing as we tweak the format. So that is really helpful. 

We need to dig a little more to find the problem though 😞

**Reply by mcanouil** [reply]
Quarto also adds plugins which might not interact well with MathJax.
It feels more and more as a JavaScript (conflict) issue.

Is there a way to opt-out Reveal.js plugins set by Quarto that I am not aware of?
